libs/langchain-openai/src/chat_models.ts

```diff
@@ -843,4 +854,118 @@ export class ChatOpenAI<
       }
     );
   }
+
+  /**
+   * Model wrapper that returns outputs formatted to match the given schema.
+   *
+   * @template {any} RunInput The input type for the Runnable.
+   * @template {z.ZodObject<any, any, any, any>} RunOutput The output type for the Runnable, expected to be a Zod schema object for structured output validation.
+   * @template {RunnableConfig} CallOptions The type for call options, extending from RunnableConfig.
+   *
+   * @param {z.ZodEffects<RunOutput>} schema The schema for the structured output. Either as a ZOD schema or a valid JSON schema object.
+   * @param {string} name The name of the function to call.
+   * @param {"functionCalling" | "jsonMode"} method The method to use for getting the structured output. Defaults to "functionCalling".
+   * @param {boolean | undefined} includeRaw Whether to include the raw output in the result. Defaults to false.
+   * @returns {Runnable<RunInput, RunOutput, CallOptions>} A new runnable that calls the LLM with structured output.
+   */
+  withStructuredOutput<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunInput = any,
+    // prettier-ignore
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    RunOutput extends z.ZodObject<any, any, any, any> = z.ZodObject<any, any, any, any>
+  >({
+    schema,
+    name,
+    method = "functionCalling",
+    includeRaw = false,
+  }: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    schema: z.ZodEffects<RunOutput> | Record<string, any>;
+    name: string;
+    method?: "functionCalling" | "jsonMode";
+    includeRaw?: boolean;
+  }): Runnable<RunInput, RunOutput> {
+    let llm: Runnable;
+    let outputParser: JsonOutputKeyToolsParser | JsonOutputParser<RunOutput>;
+
+    if (method === "jsonMode") {
+      llm = this.bind({
+        response_format: { type: "json_object" },
+      } as Partial<CallOptions>);
+      outputParser = new JsonOutputParser<RunOutput>();
+    } else {
+      // Is function calling
+      if (isZodSchema(schema)) {
+        class TmpClass extends StructuredTool {
```

Comment:
Can we just define this as JSON schema directly? This is hacky

Reply:
ya, just requires adding zodToJsonSchema as a hard dep which I wasn't sure we wanted to do.

Reply:
I think fine to

Reply:
nvm it already is!